src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -266,11 +228,22 @@ def _backchain_one_pass(
                     self._recompute_datastores_from_segments(
                         param_opt_to_nec_pnads[option.parent])
                     # Recompute all preconditions, now that we have recomputed
-                    # the datastores.
+                    # the datastores. While doing this, keep track of any
+                    # PNADs that get empty datastores.
+                    pnads_to_remove = []
                     for nec_pnad in param_opt_to_nec_pnads[option.parent]:
-                        pre = self._induce_preconditions_via_intersection(
-                            nec_pnad)
-                        nec_pnad.op = nec_pnad.op.copy_with(preconditions=pre)
+                        if len(nec_pnad.datastore) > 0:
+                            pre = self._induce_preconditions_via_intersection(
+                                nec_pnad)
+                            nec_pnad.op = nec_pnad.op.copy_with(
+                                preconditions=pre)
+                        else:
+                            pnads_to_remove.append(nec_pnad)
```

Comment:
i am curious why we didn't need to do this before, but we do now. do you understand it? is there something regarding the removal of specialization that makes it be possible now to have empty datastores during backchaining, while it wasn't possible before?

Reply:
yes - I think I understand it. So remember that when we specialize, we are only ever doing so to cover a new segment. It is thus guaranteed that that new segment will be covered by this new PNAD. However, now we're spawning a new PNAD (which will be guaranteed to cover this new segment) and recomputing datastores: which may cause a previously-existing PNAD that was covering this segment to lose all its data. This couldn't happen before because that would be exactly the PNAD that we would have specialized! (If this doesn't make sense - I'm happy to explain more via slack)

Reply:
yes makes sense, when we specialized we modified that previous PNAD, but now we're always adding a new one